### PR TITLE
[ARCTIC-788] Refactor optimize properties to self-optimizing

### LINKED ISSUE #788: [Subtask]: Change TableProperties of optimize to self-optimizing
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/arctic/issues?q=is%3Aissue) and found no similar issues.

### 结构优化配置

| 配置名称                                             | 默认值            | 描述                                               |
|-----------------------------------------------------|------------------|---------------------------------------------------|
| self-optimizing.enabled                             | true             | 是否开启结构优化                                     |
| self-optimizing.group                               | default          | 结构优化所属的组                                     |
| self-optimizing.quota                               | 0.1              | 表所能占用的结构优化资源量                             |
| self-optimizing.num-retries                         | 5                | 结构优化失败时的重试次数                               |
| self-optimizing.execute.timeout                     | 1800000（30分钟） | 结构优化执行超时时间                                  |
| self-optimizing.target-size                         | 134217728（128MB）| 结构优化的目标文件大小                                |
| self-optimizing.max-file-count                      | 100000           | 一次结构优化最多处理的文件个数                           |
| self-optimizing.max-task-file-size                  | 1073741824（1GB） | 一次结构优化最大的任务大小                          |
| self-optimizing.fragment-ratio                      | 8                | 处理的 fragment 文件阈值                             |
| self-optimizing.minor.trigger.file-count            | 12               | 触发 minor optimize 的 fragment 文件数量             |
| self-optimizing.minor.trigger.interval              | 3600000（1小时）  | 触发 minor optimize 的时间间隔                        |
| self-optimizing.major.trigger.file-count            | 12               | 触发 major optimize 的文件数量                      |
| self-optimizing.major.trigger.duplicate-ratio       | 0.5              | 触发 major optimize 的 delete 文件与目标文件大小的比值  |
| self-optimizing.major.trigger.interval              | 86400000（1天）   | 触发 major optimize 的时间间隔                        |
| self-optimizing.full.trigger.interval               | -1（关闭）         | 触发 full optimize 的时间间隔                        |

### Parent issue

https://github.com/NetEase/arctic/issues/761

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by wangtaohz**
## Why are the changes needed?

fix #788 

## Brief change log

  - refactor table properties `optimize.*` to `self-optimizing.*`
  - compatible with legacy properties
  - add new property `self-optimizing.target-size`
  - add new property `self-optimizing.fragment-ratio`
  - change `*.enable` to `*.enabled`

## How was this patch tested?
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (no)
  - If yes, how is the feature documented? (docs )

### CODE REVIEW

File: core/src/main/java/com/netease/arctic/table/TableProperties.java [hunk 1]
```diff
@@ -69,66 +69,103 @@ private TableProperties() {
   /**
    * table optimize related properties
    */
-  public static final String ENABLE_OPTIMIZE = "optimize.enable";
-  public static final String ENABLE_OPTIMIZE_DEFAULT = "true";
+  public static final String ENABLE_SELF_OPTIMIZING = "self-optimizing.enabled";
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/table/TableProperties.java:72]: Shall we change `log-store.enable` to `log-store.enabled` and `dim-table.enable` to `dim-table.enabled` too ?

Comment (wangtaohz) [reply core/src/main/java/com/netease/arctic/table/TableProperties.java:72]: Ok, I have changed them, and also the `kerberos.enabled` `kerberos.proxy.enabled` `spark.arctic.sql.delegate.enabled` `metrics.event-latency.enabled` `metrics.enabled` `log-store.consistency-guarantee.enabled` `arctic.ams.ha.enabled` `arctic.enabled`.

**Review by zhoujinsong — COMMENTED** [review]
@wangtaohz You may need to change `dist/src/main/arctic-bin/conf/config.yaml` too.

**Comment by zhoujinsong** [comment]
@YesOrNo828  @zstraw  There are some configurations changed in the Flink module.

**Comment by wangtaohz** [comment]
> 

Ok, I forgot to change the `*.yaml`.

**Comment by wangtaohz** [comment]
> @wangtaohz You may need to change `dist/src/main/arctic-bin/conf/config.yaml` too.

Ok, I forgot to change the `*.yaml`.

**Comment by zstraw** [comment]
> @YesOrNo828 @zstraw There are some configurations changed in the Flink module.

Ok, I get it.

File: core/src/main/java/com/netease/arctic/utils/CompatiblePropertyUtil.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  private static String getCompatibleProperty(Map<String, String> properties, String property) {
+    String legacyProperty = getLegacyProperty(property);
+    if (legacyProperty != null && properties.containsKey(legacyProperty) && !properties.containsKey(property)) {
+      return legacyProperty;
+    } else {
+      return property;
+    }
+  }
+
+  private static String getLegacyProperty(String property) {
```
Comment (YesOrNo828) [inline core/src/main/java/com/netease/arctic/utils/CompatiblePropertyUtil.java:68]: I think the legacy properties which are in the [ArcticValidator](https://github.com/NetEase/arctic/blob/master/flink/v1.12/flink/src/main/java/com/netease/arctic/flink/table/descriptors/ArcticValidator.java#L148) class are not compatible. e.g. `metrics.enable`, `log-store.consistency-guarantee.enable`

File: ams/ams-server/src/test/java/com/netease/arctic/ams/server/optimize/TestMajorOptimizePlan.java [hunk 3]
```diff
@@ -254,7 +254,7 @@ public void testNoPartitionTableFullOptimize() throws IOException {
         .collect(Collectors.toList()));
 
     testNoPartitionTable.updateProperties()
-        .set(TableProperties.FULL_OPTIMIZE_TRIGGER_DUPLICATE_SIZE_BYTES_THRESHOLD, "0")
+        .set(TableProperties.SELF_OPTIMIZING_MAJOR_TRIGGER_DUPLICATE_RATIO, "0.00000001")
```
Comment (YesOrNo828) [inline ams/ams-server/src/test/java/com/netease/arctic/ams/server/optimize/TestMajorOptimizePlan.java:257]: Why does the ratio change to `0.00000001`?

File: ams/ams-server/src/test/java/com/netease/arctic/ams/server/optimize/TestMajorOptimizePlan.java [hunk 4]
```diff
@@ -112,7 +112,7 @@ public void testKeyedTableFullOptimize() throws IOException {
         .collect(Collectors.toList()));
 
     testKeyedTable.updateProperties()
-        .set(TableProperties.FULL_OPTIMIZE_TRIGGER_DUPLICATE_SIZE_BYTES_THRESHOLD, "0")
+        .set(TableProperties.SELF_OPTIMIZING_MAJOR_TRIGGER_DUPLICATE_RATIO, "0.00000001")
```
Comment (YesOrNo828) [inline ams/ams-server/src/test/java/com/netease/arctic/ams/server/optimize/TestMajorOptimizePlan.java:115]: Why does the ratio change to 0.00000001?

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/IcebergFullOptimizePlan.java [hunk 5]
```diff
@@ -83,33 +84,30 @@ protected boolean partitionNeedPlan(String partitionToPath) {
...
-    long duplicateSize = PropertyUtil.propertyAsLong(arcticTable.properties(),
-        TableProperties.FULL_OPTIMIZE_TRIGGER_DUPLICATE_SIZE_BYTES_THRESHOLD,
-        TableProperties.FULL_OPTIMIZE_TRIGGER_DUPLICATE_SIZE_BYTES_THRESHOLD_DEFAULT);
-    double duplicateRatio = PropertyUtil.propertyAsDouble(arcticTable.properties(),
-        TableProperties.FULL_OPTIMIZE_TRIGGER_DUPLICATE_RATIO_THRESHOLD,
-        TableProperties.FULL_OPTIMIZE_TRIGGER_DUPLICATE_RATIO_THRESHOLD_DEFAULT);
-    // delete files total size reach target or delete files rate reach target
-    if (deleteFilesTotalSize > duplicateSize || deleteFilesTotalSize / dataFilesTotalSize >= duplicateRatio) {
+
+    long targetSize = PropertyUtil.propertyAsLong(arcticTable.properties(),
```
Comment (YesOrNo828) [inline ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/IcebergFullOptimizePlan.java:93]: Why do the `PropertyUtil` and `CompatiblePropertyUtil` work together? What's the difference?

**Review by YesOrNo828 — COMMENTED** [review]
@wangtaohz  Thanks for your contribution. I left some comments.

File: ams/ams-server/src/test/java/com/netease/arctic/ams/server/optimize/TestMajorOptimizePlan.java (hunk 4 above)
Comment (wangtaohz) [reply ams/ams-server/src/test/java/com/netease/arctic/ams/server/optimize/TestMajorOptimizePlan.java:115]: We need a small ratio to trigger `Optimizing`.  
However, `0` may be better here, I will fix it.

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/IcebergFullOptimizePlan.java (hunk 5 above)
Comment (wangtaohz) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/IcebergFullOptimizePlan.java:93]: I used them together with carelessness, although It's safe to use `CompatiblePropertyUtil` here. I'll fix it.

If a new property is not compatible with any legacy property, there is no need to use `CompatiblePropertyUtil`.

File: ams/ams-server/src/test/java/com/netease/arctic/ams/server/optimize/TestMajorOptimizePlan.java (hunk 3 above)
Comment (wangtaohz) [reply ams/ams-server/src/test/java/com/netease/arctic/ams/server/optimize/TestMajorOptimizePlan.java:257]: 0 may be better here, I will fix it

File: core/src/main/java/com/netease/arctic/utils/CompatiblePropertyUtil.java (hunk 2 above)
Comment (wangtaohz) [reply core/src/main/java/com/netease/arctic/utils/CompatiblePropertyUtil.java:68]: Ok, I will check and add the compatible use of properties in [ArcticValidator](https://github.com/NetEase/arctic/blob/master/flink/v1.12/flink/src/main/java/com/netease/arctic/flink/table/descriptors/ArcticValidator.java#L148).

But, does it have to be compatible with legacy `log-store.consistency-guarantee.enable`? It seems to be a session property, the same as `dim-table.enable`.

Comment (wangtaohz) [reply core/src/main/java/com/netease/arctic/utils/CompatiblePropertyUtil.java:68]: I will open another PR to be compatible with `log-store.consistency-guarantee.enable` and `dim-table.enable`.

**Review by zhoujinsong — APPROVED** [review]
LGTM.
